applications/FluidDynamicsApplication/tests/cpp_tests/test_shock_detection_process.cpp

```diff
@@ -0,0 +1,165 @@
+                f_r = std::abs(std::sin(2.0 * Globals::Pi * r));
+            } else if (r >= one_third) {
+                f_r = 2.0 * r - 1 + (1.0 / 6.0) * std::sin(3.0 * Globals::Pi * r);
+            } else {
+                f_r = - r * std::sin(1.5 * Globals::Pi * std::pow(r,2));
+            }
+            r_rho = x_check ? f_r + std::cos(2.0 * Globals::Pi * r_node.Y()) : f_r;
+        }
+    }
+
+    /**
+     * Checks the shock detection process with a smooth field with no expected shocks
+     */
+    KRATOS_TEST_CASE_IN_SUITE(ShockDetectionSmoothField, FluidDynamicsApplicationFastSuite)
+    {
+        // Set the test model part
+        Model model;
+        auto& r_model_part = model.CreateModelPart("MainModelPart");
+        SetTestModelPart(r_model_part);
+
+        // Set a smooth field
+        for (auto& r_node : r_model_part.Nodes()) {
+            r_node.FastGetSolutionStepValue(DENSITY) = r_node.X() * r_node.Y();
+        }
+
+        // Perform the shock detection
+        ShockDetectionProcess shock_detection(r_model_part);
+        shock_detection.ExecuteInitialize();
+        shock_detection.EdgeBasedShockDetection(DENSITY, DENSITY_GRADIENT);
+
+        // Check values
+        const double tolerance = 1.0e-8;
+        KRATOS_CHECK_NEAR(r_model_part.GetNode(5100).GetValue(SHOCK_SENSOR), 0.0, tolerance);
+
+        // GidIO<> gid_io_abgrall(
+        //     "/home/rzorrilla/Desktop/abgrall_function_smooth_field",
+        //     GiD_PostAscii,
+        //     SingleFile,
+        //     WriteDeformed,
+        //     WriteConditions);
+		// gid_io_abgrall.InitializeMesh(0.0);
+		// gid_io_abgrall.WriteMesh(r_model_part.GetMesh());
+		// gid_io_abgrall.FinalizeMesh();
+		// gid_io_abgrall.InitializeResults(0, r_model_part.GetMesh());
+		// gid_io_abgrall.WriteNodalResults(DENSITY, r_model_part.Nodes(), 0, 0);
+		// gid_io_abgrall.WriteNodalResultsNonHistorical(NODAL_AREA, r_model_part.Nodes(), 0);
+		// gid_io_abgrall.WriteNodalResultsNonHistorical(SHOCK_SENSOR, r_model_part.Nodes(), 0);
+		// gid_io_abgrall.WriteNodalResultsNonHistorical(DENSITY_GRADIENT, r_model_part.Nodes(), 0);
+		// gid_io_abgrall.FinalizeResults();
+
+    }
+
+    /**
+     * Checks the shock detection process with the Abgrall function
+     */
+    KRATOS_TEST_CASE_IN_SUITE(ShockDetectionAbgrallFunction, FluidDynamicsApplicationFastSuite)
+    {
+        // Set the test model part
+        Model model;
+        auto& r_model_part = model.CreateModelPart("MainModelPart");
+        SetTestModelPart(r_model_part);
+        SetAbgrallFunction(r_model_part);
+
+        // Perform the shock detection
+        ShockDetectionProcess shock_detection(r_model_part);
+        shock_detection.ExecuteInitialize();
+        shock_detection.EdgeBasedShockDetection(DENSITY, DENSITY_GRADIENT);
+
+        // Check values
+        const double tolerance = 1.0e-6;
+        KRATOS_CHECK_NEAR(r_model_part.GetNode(4223).GetValue(SHOCK_SENSOR), 1.0, tolerance);
```

Comment:
just to tell that if you have this way of testing, doing the test also in mpi would be very easy

Reply:
We first need the MPI implementation of the `ComputeNodalGradientProcess`. I'd do it in another PR.